### ISSUE TITLE
Allow a control flow statement to be decorated with #[allow(unreachable_code)]

### ISSUE THREAD

**Issue by mqudsi** [rust-lang/rust#80671]
Currently it seems that the only way to silence an `unreachable_code` warning is via an interior `#![allow(unreachable_code)]` attribute that would apply to the entire scope or artificially creating a new scope with the inner attribute in question.

Would it be possible to allow a control flow statement to be individually tagged with a regular `#[allow(unreachable_code)]` statement to also suffice?

e.g. the following should not emit a warning about unreachable code:

```rust
pub fn foo() -> u32 {
    #[allow(unreachable_code)]
    return 0;

    let foo = 7;
    return foo;
}
```

[playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&code=pub%20fn%20foo()%20-%3E%20u32%20%7B%0A%20%20%20%20%23%5Ballow(unreachable_code)%5D%0A%20%20%20%20return%200%3B%0A%0A%20%20%20%20let%20foo%20%3D%207%3B%0A%20%20%20%20return%20foo%3B%0A%7D%0A)

**Comment by oli-obk** [comment]
You'd need to apply the attribute to the `let`, not the `return`, as the `return` is not unreachable, it's everything behind it.

We frequently have similar issues in clippy with diagnostics reporting on an expression, but lint attributes only working on entire functions. I'm not sure if the `unreachable_code` can be fixed in the way it is implemented right now, but there's nothing inherently speaking against making it more fine grained.

**Comment by mqudsi** [comment]
I originally had an example showing both the `let` and the `return` each decorated with `#[allow(unreachable_code)]` but decided to make the feature request more specific (as neither suppresses the warning right now). I figured that semantically a single unreachable expression tagged with `allow(unreachable_code)` would, strictly speaking, apply only to that one expression and nothing would explicitly stop the compiler from complaining about the line that came after... but then again, what contributes to making one or more statements unreachable could also be a combination of control flow points (e.g. multiple if statements covering all the cases) in which case it's not much better either.

If interior attributes not at the start of a block/section were allowed to mean "from here on out" that would probably come closest to conveying the intended semantics (and I could see plenty of other uses for it) but that's quite a departure from the core issue at hand.
